src/modules/musig/main_impl.h

```diff
@@ -0,0 +1,586 @@
+    size_t i;
+
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(session != NULL);
+    ARG_CHECK(signers != NULL);
+    ARG_CHECK(combined_pk != NULL);
+    ARG_CHECK(pk_hash32 != NULL);
+    ARG_CHECK(pubkeys != NULL);
+    ARG_CHECK(commitments != NULL);
+    for (i = 0; i < n_signers; i++) {
+        ARG_CHECK(commitments[i] != NULL);
+    }
+    (void) ctx;
+
+    memcpy(&session->combined_pk, combined_pk, sizeof(*combined_pk));
+    if (n_signers == 0) {
+        return 0;
+    }
+    session->n_signers = n_signers;
+    secp256k1_musig_signers_init(signers, pubkeys, n_signers);
+
+    memcpy(session->pk_hash, pk_hash32, 32);
+    session->nonce_is_set = 0;
+    session->msg_is_set = 0;
+    if (msg32 != NULL) {
+        memcpy(session->msg, msg32, 32);
+        session->msg_is_set = 1;
+    }
+    session->has_secret_data = 0;
+
+    for (i = 0; i < n_signers; i++) {
+        memcpy(signers[i].nonce_commitment, commitments[i], 32);
+    }
+    return 1;
+}
+
+int secp256k1_musig_set_nonce(const secp256k1_context* ctx, secp256k1_musig_session_signer_data *signer, const secp256k1_pubkey *nonce) {
+    unsigned char commit[33];
+    size_t commit_size = sizeof(commit);
+    secp256k1_sha256 sha;
+
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(signer != NULL);
+    ARG_CHECK(nonce != NULL);
+
+    secp256k1_sha256_initialize(&sha);
+    secp256k1_ec_pubkey_serialize(ctx, commit, &commit_size, nonce, SECP256K1_EC_COMPRESSED);
+    secp256k1_sha256_write(&sha, commit, commit_size);
+    secp256k1_sha256_finalize(&sha, commit);
+
+    if (memcmp(commit, signer->nonce_commitment, 32) != 0) {
+        return 0;
+    }
+    memcpy(&signer->nonce, nonce, sizeof(*nonce));
+    signer->present = 1;
+    return 1;
+}
+
+int secp256k1_musig_session_combine_nonces(const secp256k1_context* ctx, secp256k1_musig_session *session, const secp256k1_musig_session_signer_data *signers, size_t n_signers, int *nonce_is_negated, const secp256k1_pubkey *adaptor) {
+    secp256k1_gej combined_noncej;
+    secp256k1_ge combined_noncep;
+    secp256k1_ge noncep;
+    secp256k1_sha256 sha;
+    unsigned char nonce_commitments_hash[32];
+    size_t i;
+
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(session != NULL);
+    ARG_CHECK(signers != NULL);
+
+    if (n_signers != session->n_signers) {
+        return 0;
+    }
+    secp256k1_sha256_initialize(&sha);
+    secp256k1_gej_set_infinity(&combined_noncej);
+    for (i = 0; i < n_signers; i++) {
+        if (!signers[i].present) {
+            return 0;
+        }
+        secp256k1_sha256_write(&sha, signers[i].nonce_commitment, 32);
+        secp256k1_pubkey_load(ctx, &noncep, &signers[i].nonce);
+        secp256k1_gej_add_ge_var(&combined_noncej, &combined_noncej, &noncep, NULL);
+    }
+    secp256k1_sha256_finalize(&sha, nonce_commitments_hash);
+    if (session->has_secret_data
+            && memcmp(session->nonce_commitments_hash, nonce_commitments_hash, 32) != 0) {
```

Comment:
We should access `session->nonce_commitments_hash` only if `     session->nonce_commitments_hash_is_set` (undefined behavior otherwise)

Reply:
Maybe you actually wanted to test for that and not for `has_secret_data`?

Reply:
maybe (VERIFY_)CHECK it

Reply:
adding VERIFY_CHECK